### PR TITLE
Add Tests to Template Tests

### LINKED ISSUE #29: Add JUnit Testing
https://nus-cs2113-ay2021s1.github.io/website/schedule/week8/project.html#4-add-some-junit-tests

### PR DESCRIPTION
**Pull request by PraveenElango**
Included testing code for templates

### CODE REVIEW

File: src/main/java/seedu/duke/template/LcTemplate.java [hunk 1]
```diff
@@ -7,12 +7,12 @@
...
 
-    private static final String LC_TEMPLATE = "\t+---L-----C---+\n"
+    protected static final String LC_TEMPLATE = "\t+---L-----C---+\n"
                                                 + "\t|             |\n"
                                                 + "\t|             |\n"
                                                 + "\t+----+V_ac+---+\n";
-    private Inductor inductor;
-    private Capacitor capacitor;
+    protected Inductor inductor;
+    protected Capacitor capacitor;
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/template/LcTemplate.java]: What's the reasoning for making this (as well as in the other classes) `protected` instead of `private`?

File: src/test/java/seedu/duke/template/LrTemplateTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package seedu.duke.template;
+
+import org.junit.jupiter.api.Test;
+import seedu.duke.DukeException;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class LrTemplateTest {
+    public static final double DELTA = -5 * 10 ^ -5;
```
Comment (hughjazzman) [inline src/test/java/seedu/duke/template/LrTemplateTest.java]: Probably don't want to XOR here right?
```suggestion
    public static final double DELTA = 1e-15;
```

File: src/test/java/seedu/duke/template/LcTemplateTest.java [hunk 3]
```diff
@@ -2,14 +2,109 @@
...
+
+    @Test
+    void setComponent_setCapacitorValue() {
+        LcTemplate t = new LcTemplate();
+        t.setComponent("c", 5);
+        assertEquals(5, t.getCapacitor().getValue(), DELTA);
+    }
+
+    @Test
+    void toString_returnsString() {
```
Comment (hughjazzman) [inline src/test/java/seedu/duke/template/LcTemplateTest.java]: Thinking the return toString methods could simply be tested in text-ui-test. Kind of odd that we just copy paste for JUnit.

File: src/test/java/seedu/duke/template/LcTemplateTest.java [hunk 4]
```diff
@@ -2,14 +2,109 @@
...
+    public static final double DELTA = 1e-15;
 
     @Test
     void calcImpedance_zeroCapacitance_expectException() {
         LcTemplate t = new LcTemplate();
         assertThrows(DukeException.class, t::calcImpedance);
     }
+
+    @Test
+    void calcImpedance_returnsDouble() throws DukeException {
```
Comment (hughjazzman) [inline src/test/java/seedu/duke/template/LcTemplateTest.java]: Consider changing all method names to include the test input/scenario.
https://se-education.org/guides/conventions/java/basic.html#naming

File: src/test/java/seedu/duke/template/LcTemplateTest.java [hunk 5]
```diff
@@ -2,14 +2,109 @@
...
+        LcTemplate t = new LcTemplate(1,1,1);
+        double inductance = 1 * Math.pow(10, -6);
+        double capacitance = 1 * Math.pow(10, -6);
+        double impedance = t.ANGULAR_FREQUENCY * Math.abs(inductance - (1 / capacitance));
+        assertEquals(impedance, t.calcImpedance(), DELTA);
+    }
+
+    @Test
+    void getInductor_returnsInductor() {
```
Comment (hughjazzman) [inline src/test/java/seedu/duke/template/LcTemplateTest.java]: Consider adding edge cases? Like 0, or negative values, to make the code more robust - similar for all the methods.

File: src/test/java/seedu/duke/template/LcTemplateTest.java [hunk 6]
```diff
@@ -2,14 +2,109 @@
...
+
+    @Test
+    void getComponent_getInductorValue() throws DukeException {
+        LcTemplate t = new LcTemplate();
+        t.setComponent("l", 5);
+        assertEquals(5, t.getComponent("l").getValue(), DELTA);
+    }
+
+    @Test
+    void getComponent_expectException() {
```
Comment (hughjazzman) [inline src/test/java/seedu/duke/template/LcTemplateTest.java]: i.e. test scenario here could be `noValue`

**Review by hughjazzman — CHANGES_REQUESTED** [review]
Some comments to address, but looks good overall!

File: src/main/java/seedu/duke/template/LcTemplate.java (hunk 1 above)
Comment (PraveenElango) [reply src/main/java/seedu/duke/template/LcTemplate.java]: I was not able to access them in the test files where I was trying to reference them. Changing from private to protected solved that problem.

File: src/test/java/seedu/duke/template/LrTemplateTest.java (hunk 2 above)
Comment (PraveenElango) [reply src/test/java/seedu/duke/template/LrTemplateTest.java]: Oops, my bad! I'll use the pow function instead

File: src/main/java/seedu/duke/template/LcTemplate.java (hunk 1 above)
Comment (hughjazzman) [reply src/main/java/seedu/duke/template/LcTemplate.java]: Ah okay, but I would say there is no need to access the `TEMPLATE` (see comment about the `toString`), and the components could be accessed by getters instead, so actually don't need to be `protected`?

File: src/test/java/seedu/duke/template/LcTemplateTest.java (hunk 3 above)
Comment (PraveenElango) [reply src/test/java/seedu/duke/template/LcTemplateTest.java]: Sure, that sounds good.

File: src/test/java/seedu/duke/template/LcTemplateTest.java (hunk 4 above)
Comment (PraveenElango) [reply src/test/java/seedu/duke/template/LcTemplateTest.java]: Got it, will rename accordingly!

File: src/test/java/seedu/duke/template/LcTemplateTest.java (hunk 5 above)
Comment (PraveenElango) [reply src/test/java/seedu/duke/template/LcTemplateTest.java]: Right, I'll figure that out. Thanks Wira :)

File: src/test/java/seedu/duke/template/LcTemplateTest.java (hunk 6 above)
Comment (PraveenElango) [reply src/test/java/seedu/duke/template/LcTemplateTest.java]: Agreed, will change.

File: src/main/java/seedu/duke/template/LcTemplate.java (hunk 1 above)
Comment (PraveenElango) [reply src/main/java/seedu/duke/template/LcTemplate.java]: Oh, but aren't we testing the getter functions inside the templates and not the components?

Comment (hughjazzman) [reply src/main/java/seedu/duke/template/LcTemplate.java]: Trivial getters wouldn't really need testing in the end: https://stackoverflow.com/questions/6197370/should-unit-tests-be-written-for-getter-and-setters

**Review by hughjazzman — APPROVED** [review]
LGTM
